### PR TITLE
feat(c++): recover the BUILD_ARROW_FROM_SOURCE option

### PR DESCRIPTION
**Pull request by acezen**
### Reason for this PR

There may be senerio that user want to build graphar as a whole static library that include all the dependencies. arrow's static distribution does not include some libraries like thrift, aws...
So we may build arrow from the source.

### What changes are included in this PR?

- recover the `BUILD_ARROW_FROM_SOURCE` to support build arrow from source and link graphar as a single static library that includes all the dependencies.
- add a nightly ci to check the option.

### Are these changes tested?

yes
### Are there any user-facing changes?

no

### CODE REVIEW

File: cpp/cmake/apache-arrow.cmake [hunk 1]
```diff
@@ -107,11 +108,11 @@ function(build_arrow)
...
-    if (Arrow_FOUND) # arrow is installed, build the same version as the installed one
-        message(STATUS "Found Arrow installed, align to version: ${Arrow_VERSION}")
-        set(ARROW_VERSION_TO_BUILD "${Arrow_VERSION}" CACHE INTERNAL "arrow version")
-    endif ()
+    set(ARROW_VERSION_TO_BUILD "15.0.0" CACHE INTERNAL "arrow version")
+    # if (Arrow_FOUND) # arrow is installed, build the same version as the installed one
+        # message(STATUS "Found Arrow installed, align to version: ${Arrow_VERSION}")
+        # set(ARROW_VERSION_TO_BUILD "${Arrow_VERSION}" CACHE INTERNAL "arrow version")
+    # endif ()
     set(GAR_ARROW_SOURCE_FILE "https://www.apache.org/dyn/closer.lua?action=download&filename=arrow/arrow-${ARROW_VERSION_TO_BUILD}/apache-arrow-${ARROW_VERSION_TO_BUILD}.tar.gz")
```
Comment (jasinliu) [inline cpp/cmake/apache-arrow.cmake:116]: Can environment variables or parameters be passed here?

Comment (jasinliu) [reply cpp/cmake/apache-arrow.cmake:116]: If the user has their own download link, they can identify and determine the version after installation.

> Can environment variables or parameters be passed here?

Comment (acezen) [reply cpp/cmake/apache-arrow.cmake:116]: hi, @jasinliu, thanks for the advice. Maybe we can not expose the variable to user, and if there is arrow installed in the system, we can align to the version

Comment (acezen) [reply cpp/cmake/apache-arrow.cmake:116]: > If the user has their own download link, they can identify and determine the version after installation.
> 
> > Can environment variables or parameters be passed here?

and I not sure what you mean own download link? you mean there maybe arrow that modified by user?

Comment (jasinliu) [reply cpp/cmake/apache-arrow.cmake:116]: > > If the user has their own download link, they can identify and determine the version after installation.
> > > Can environment variables or parameters be passed here?
> 
> and I not sure what you mean own download link? you mean there maybe arrow that modified by user?

For example, the specified mirror link， the modification of the arrow，or some test version of the arrow. Custom links can provide some flexibility。

Comment (acezen) [reply cpp/cmake/apache-arrow.cmake:116]: > > > If the user has their own download link, they can identify and determine the version after installation.
> > > > Can environment variables or parameters be passed here?
> > 
> > 
> > and I not sure what you mean own download link? you mean there maybe arrow that modified by user?
> 
> For example, the specified mirror link， the modification of the arrow，or some test version of the arrow. Custom links can provide some flexibility。

But we are supposed to only support the original version of arrow, provide a link for that is a little weird. I suggest we still use the link from apache, and if user want to use other link, he could modify the cmake file directly.

**Review by lixueclaire — APPROVED** [review]
LGTM

**Pull request merged**